### ISSUE TITLE
Rule ids convertible to integer are cast to floats

### ISSUE THREAD

**Issue by crazyfermions** [semgrep/semgrep#4955]
**Describe the bug**
Test execution fails for rules
1. if the rule id is convertible to integer, e.g. '1023225'
2.  AND if they are in a subdirectory which also happens to be compatible to be an integer as well (e.g. also '1023225').

The rule is apparently parsed to `<folder>.<ruleid>`, interpreted as a float and then rounded resulting into a key error, e.g. here '1023225.1032'. The bug can be avoided if the rule id or the folder name is changed to contain other characters, such as 'a-1023225'

**To Reproduce**
Install semgrep 0.86.5 via pypi. Run a rule file with an ID convertible to integer, e.g. '1023225' inside a folder also converitble to an integer, e.g. 

```
semgrep -c 1023225/semgrep.yaml src/
```

This does *not* occur on https://semgrep.dev (probably because 2. is not given)

**Expected behavior**
Rules are executed normally.

**Screenshots**

```
Traceback (most recent call last):
  File "[REDACTED]/venv/bin/semgrep", line 8, in <module>
    sys.exit(main())
  File "[REDACTED]/venv/lib/python3.8/site-packages/semgrep/__main__.py", line 20, in main
    cli()
  File "[REDACTED]/venv/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "[REDACTED]/venv/lib/python3.8/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "[REDACTED]/venv/lib/python3.8/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "[REDACTED]/venv/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "[REDACTED]/venv/lib/python3.8/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "[REDACTED]/venv/lib/python3.8/site-packages/semgrep/commands/scan.py", line 868, in scan
    ) = semgrep.semgrep_main.main(
  File "[REDACTED]/venv/lib/python3.8/site-packages/semgrep/semgrep_main.py", line 340, in main
    rule_matches_by_rule, semgrep_errors, all_targets, profiling_data = run_rules(
  File "[REDACTED]/venv/lib/python3.8/site-packages/semgrep/semgrep_main.py", line 146, in run_rules
    ) = core_runner.invoke_semgrep(
  File "[REDACTED]/venv/lib/python3.8/site-packages/semgrep/core_runner.py", line 680, in invoke_semgrep
    ) = self._run_rules_direct_to_semgrep_core(
  File "[REDACTED]/venv/lib/python3.8/site-packages/semgrep/core_runner.py", line 636, in _run_rules_direct_to_semgrep_core
    core_output = CoreOutput.parse(rules, output_json)
  File "[REDACTED]/venv/lib/python3.8/site-packages/semgrep/core_output.py", line 292, in parse
    parsed_timings = CoreTiming.parse(rule_table, timings)
  File "[REDACTED]/venv/lib/python3.8/site-packages/semgrep/core_output.py", line 251, in parse
    rules = [rule_table[RuleId(rule)] for rule in raw_json.get("rules", [])]
  File "[REDACTED]/venv/lib/python3.8/site-packages/semgrep/core_output.py", line 251, in <listcomp>
    rules = [rule_table[RuleId(rule)] for rule in raw_json.get("rules", [])]
KeyError: '1023225.10232'
```

I dug with pdb and found out that the rule id is sent correctly to the core, but at some point in the core it breaks:
```
[REDACTED]/venv/lib/python3.8/site-packages/semgrep/core_runner.py(636)_run_rules_direct_to_semgrep_core()
(Pdb) rules
[<Rule id=1023225.1023225>]
(Pdb) runner.stdout
'{"matches":[],"errors":[{"error_type":"Fatal error","severity":"error","location":{"path":"[REDACTED]","start":{"line":1,"col":1,"offset":0},"end":{"line":1,"col":1,"offset":0},"lines":["[REDACTED]"]},"message":"Not_found","details":"Raised at Stdlib__hashtbl.find in file \\"hashtbl.ml\\", line 539, characters 13-28\\nCalled from Stdlib__list.map in file \\"list.ml\\", line 92, characters 20-23\\nCalled from Run_semgrep.semgrep_with_rules.(fun) in file \\"src/runner/Run_semgrep.ml\\", line 474, characters 13-113\\nCalled from Run_semgrep.iter_targets_and_get_matches_and_exn_to_errors.(fun) in file \\"src/runner/Run_semgrep.ml\\", line 305, characters 21-29\\nCalled from Stdlib__fun.protect in file \\"fun.ml\\", line 33, characters 8-15\\nRe-raised at Stdlib__fun.protect in file \\"fun.ml\\", line 38, characters 6-52\\nCalled from Memory_limit.run_with_memory_limit in file \\"src/system/Memory_limit.ml\\", line 77, characters 6-62\\nCalled from Run_semgrep.iter_targets_and_get_matches_and_exn_to_errors.(fun) in file \\"src/runner/Run_semgrep.ml\\", line 296, characters 17-1023\\n"}],"skipped":[],"stats":{"okfiles":0,"errorfiles":1},"time":{"targets":[{"path":[REDACTED]","rule_times":[],"run_time":0.00014495849609375}],"rules":["1023225.10232"],"rules_parse_time":0.00475001335144043}}\n'
```

**What is the priority of the bug to you?**

- [ ] P0: blocking your adoption of Semgrep or workflow
- [x] P1: important to fix or quite annoying
- [ ] P2: regular bug that should get fixed

**Environment**
Py 3.8 + semgrep 0.86.5 via pypi, but it also occurs with previous versions.

**Use case**
Rules anf folders can have allowed identifiers.

UPDATE: Found out that this also requires the folder to be a valid integer name, updated accordingly.

**Comment by r2c-demo** [comment]
This issue is synced in Linear at https://linear.app/r2c/issue/PA-1104/rule-ids-convertible-to-integer-are-cast-to-floats.

**Comment by ievans** [comment]
Great bug report @crazyfermions, thank you!

**Comment by aryx** [comment]
@mjambon could this issue be related to the recent fix you did in pfff about JSON Int and Float?

**Comment by mjambon** [comment]
I suspect this is an issue with YAML syntax. Could we see the YAML file? If the rule name is `3.14`, it needs to be quoted so that it's interpreted as a string and not as a number. In doubt, use JSON syntax since it's compatible with YAML and has sane escaping rules.

Do:
```yaml
rules:
- id: "3.14"
```
Don't:
```yaml
rules:
- id: 3.14
```

**Comment by crazyfermions** [comment]
@mjambon Minimal example:

```yaml
rules:
- id: "123456789"
  languages:
  - "C"
  message: "Rule text"
  pattern-either:
        - pattern: "p1"
        - pattern: "p2"
  severity: "WARNING"
```

Place into a folder `123456789/semgrep.yaml`, call semgrep with `semgrep -c 123456789/semgrep.yaml`, receive `KeyError: '123456789.124'`.

**Comment by mjambon** [comment]
I'm not getting an error with the development version of semgrep. Maybe this was fixed recently. The latest version of semgrep is 0.97.0.

```
$ semgrep -c 123456789/semgrep.yaml <(echo 'p1;')
Scanning 1 file.

Findings:

  /tmp/tmpbg6au09c/dev_fd_63 
     123456789.123456789
        Rule text

          1┆ p1;

Some files were skipped or only partially analyzed.
  Scan was limited to files tracked by git.

Ran 1 rule on 1 file: 1 finding.

A new version of Semgrep is available. See https://semgrep.dev/docs/upgrading
```

```
$ cat 123456789/semgrep.yaml
rules:
- id: "123456789"
  languages:
  - C
  message: "Rule text"
  pattern-either:
        - pattern: "p1"
        - pattern: "p2"
  severity: "WARNING"
```

**Comment by crazyfermions** [comment]
I can confirm that this issue was fixed with 0.91 (i.e. it occurs in 0.90 but no longer in 0.91).
